core/trie/proof.go

```diff
@@ -172,8 +194,272 @@ func VerifyProof(root *felt.Felt, key *Key, value *felt.Felt, proofs []ProofNode
+		}
+
+		lastProofPath, err = ProofToPath(proofs[1], proofKeys[1], hash)
+		if err != nil {
+			return false, err
+		}
+	}
+
+	// Step 2: Build trie from proofPaths and keys
+	tmpTrie, err := BuildTrie(firstProofPath, lastProofPath, keys, values)
+	if err != nil {
+		return false, err
+	}
+
+	// Verify that the recomputed root hash matches the provided root hash
+	recomputedRoot, err := tmpTrie.Root()
+	if err != nil {
+		return false, err
+	}
+	if !recomputedRoot.Equal(root) {
+		return false, errors.New("root hash mismatch")
+	}
+
+	return true, nil
+}
+
+func ensureMonotonicIncreasing(proofKeys [2]*felt.Felt, keys []*felt.Felt) error {
+	if proofKeys[0] != nil && proofKeys[0].Cmp(keys[0]) >= 0 {
+		return errors.New("range is not monotonically increasing")
+	}
+	if proofKeys[1] != nil && keys[len(keys)-1].Cmp(proofKeys[1]) >= 0 {
+		return errors.New("range is not monotonically increasing")
+	}
+	if len(keys) > 2 {
+		for i := range keys[0 : len(keys)-2] {
+			if keys[i].Cmp(keys[i+1]) >= 0 {
+				return errors.New("range is not monotonically increasing")
+			}
+		}
+	}
+	return nil
+}
+
+// shouldSquish determines if the node needs compressed, and if so, the len needed to arrive at the next key
+func shouldSquish(idx int, proofNodes []ProofNode) (int, uint8) {
+	parent := &proofNodes[idx]
+	var child *ProofNode
+	// The child is nil of the current node is a leaf
+	if idx != len(proofNodes)-1 {
+		child = &proofNodes[idx+1]
+	}
+
+	if child == nil {
+		return 0, 0
+	}
+
+	if parent.Edge != nil && child.Binary != nil {
+		return 1, parent.Edge.Path.len
+	}
+
+	if parent.Binary != nil && child.Edge != nil {
+		return 1, child.Edge.Path.len
+	}
+
+	return 0, 0
+}
+
+// ProofToPath returns the set of storage nodes along the proofNodes towards the leaf.
+// Note that only the nodes and children along this path will be set correctly.
+func ProofToPath(proofNodes []ProofNode, leaf *felt.Felt, hashF hashFunc) ([]StorageNode, error) {
+	pathNodes := []StorageNode{}
+	leafBytes := leaf.Bytes()
+	leafKey := NewKey(251, leafBytes[:]) //nolint:gomnd
+
+	// Hack: this allows us to store a right without an existing left node.
+	zeroFeltBytes := new(felt.Felt).SetUint64(0).Bytes()
+	nilKey := NewKey(0, zeroFeltBytes[:])
+
+	i, offset := 0, 0
+	for i <= len(proofNodes)-1 {
+		var crntKey *Key
+		crntNode := Node{}
+
+		height := getHeight(i, pathNodes, proofNodes)
+
+		// Set the key of the current node
+		squishedParent, squishParentOffset := shouldSquish(i, proofNodes)
+		if proofNodes[i].Binary != nil {
+			crntKey = leafKey.SubKey(height)
+		} else {
+			crntKey = leafKey.SubKey(height + squishParentOffset)
+		}
+		offset += squishedParent
+
+		// Set the value of the current node
+		crntNode.Value = proofNodes[i].Hash(hashF)
+
+		// Set the children of the current node
+		// The child may need compressed. If so, point to its compressed form.
+		childIdx := i + squishedParent + 1
+		if i+2+squishedParent < len(proofNodes)-1 {
```

Comment:
This if-else here is pure magic - I cannot understand why we're looking at 3 different cases:
- `i+2+squishedParent < len(proofNodes)-1`
- `i+1+offset == len(proofNodes)-1`
- otherwise

also I feel me might want to refactor the `crntNode` construction into a functions which would do the `leafKey.Test(...)` because a lot of code is duplicated or similar and it's hard to follow what happens

Reply:
I added some comments and refactored a bit. Let me know if it's still unclear!

Reply:
Just an update: this logic is / will be simplified in [this](https://github.com/NethermindEth/juno/pull/1901/files) PR

Reply:
No need to push it here, we can review the changes in PR #1901